Document session variables related to full scans. (#14590)
cockroachdb/docs: v21.2/cost-based-optimizer.md

When `sql.stats.histogram_collection.enabled` is set to `false`, histograms are never collected, either as part of automatic statistics collection or by manually invoking [`CREATE STATISTICS`](create-statistics.html).

## Control whether the optimizer creates a plan with a full scan

> Even if you have [secondary indexes](schema-design-indexes.html), the optimizer may decide that a full table scan will be faster. For example, if you add a secondary index to a table with a large number of rows and determine that a statement plan isn't using the secondary index, it is likely that performing a full table scan using the primary key is faster than doing a secondary index scan plus an [index join](indexes.html#example).

Reviewer: ```suggestion
Even if you have [secondary indexes](schema-design-indexes.html), the optimizer may determine that a full table scan will be faster. For example, if you add a secondary index to a table with a large number of rows and find that a statement plan is not using the secondary index, then it is likely that performing a full table scan using the primary key is faster than doing a secondary index scan plus an [index join](indexes.html#example).
```

Reviewer: I was slightly slow to parse the second sentence, so am making these suggestions for readability. Feel free to disregard. However, I feel that "determine" is better than "decide" because the latter can imply (at least to me) that the optimizer's choice might be arbitrary and/or incorrect.

You can disable statement plans that perform full table scans with the `disallow_full_table_scans` [session variable](set-vars.html).

If you disable full scans, you can set the `large_full_scan_rows` session variable to specify the maximum table size allowed for a full scan. If no alternative plan is possible, the optimizer will return an error.

> If you disable full scans, and you provide an [index hint](table-expressions.html#force-index-selection), the optimizer will try to avoid a full scan while also respecting the index hint. If this is not possible, the optimizer will return an error. If you don't provide an index hint, the optimizer will return an error, the full scan will be logged, and the `sql.guardrails.full_scan_rejected.count` [metric](ui-overview-dashboard.html) will be updated.

Reviewer: ```suggestion
If you disable full scans, and you provide an [index hint](table-expressions.html#force-index-selection), the optimizer will try to avoid a full scan while also respecting the index hint. If this is not possible, the optimizer will return an error. If you do not provide an index hint, the optimizer will return an error, the full scan will be logged, and the `sql.guardrails.full_scan_rejected.count` [metric](ui-overview-dashboard.html) will be updated.
```
